### PR TITLE
Replace parse5 and nwmatcher by jsdom

### LINKED ISSUE #9: Error: The string "", is not a valid CSS selector
Hi, 
I am using the `CrittersPlugin` with the `OptimizeCssAssetsPlugin`. When I add this the fallowing css 
```css
* {
    box-sizing: border-box;
}

*::before,
*::after {
    box-sizing: inherit;
}
``` 
it gets minified to:
```css
*{box-sizing:border-box}:after,:before{box-sizing:inherit}
```
Notice it is now missing the `*` selector and it only uses `:after` and `:before`.

If I remove the css rules for `*::before` and `*::after` or if I `CrittersPlugin`  from the plugins array it works fine, but when I include them I get the fallowing error:

```
ERROR in   Error: The string "", is not a valid CSS selector

  - nwmatcher.js:827 emit
    [webpack-tutorial-boilerplate]/[nwmatcher]/src/nwmatcher.js:827:37

  - nwmatcher.js:1529 select
    [webpack-tutorial-boilerplate]/[nwmatcher]/src/nwmatcher.js:1529:11

  - nwmatcher.js:1451 Object.first
    [webpack-tutorial-boilerplate]/[nwmatcher]/src/nwmatcher.js:1451:14

  - critters.js:117 Node.querySelector
    [webpack-tutorial-boilerplate]/[critters-webpack-plugin]/dist/critters.js:117:28

  - critters.js:359
    [webpack-tutorial-boilerplate]/[critters-webpack-plugin]/dist/critters.js:359:37

  - Array.filter

  - critters.js:357
    [webpack-tutorial-boilerplate]/[critters-webpack-plugin]/dist/critters.js:357:49

  - critters.js:166
    [webpack-tutorial-boilerplate]/[critters-webpack-plugin]/dist/critters.js:166:16

  - Array.filter

  - critters.js:162 walkStyleRules
    [webpack-tutorial-boilerplate]/[critters-webpack-plugin]/dist/critters.js:162:29

  - critters.js:161 walkStyleRules
    [webpack-tutorial-boilerplate]/[critters-webpack-plugin]/dist/critters.js:161:18

  - critters.js:355 Critters.<anonymous>
    [webpack-tutorial-boilerplate]/[critters-webpack-plugin]/dist/critters.js:355:9

  - new Promise

  - critters.js:345 Critters.processStyle
    [webpack-tutorial-boilerplate]/[critters-webpack-plugin]/dist/critters.js:345:12

  - critters.js:254
    [webpack-tutorial-boilerplate]/[critters-webpack-plugin]/dist/critters.js:254:76

  - Array.map

  - critters.js:254 Critters.$If_2
    [webpack-tutorial-boilerplate]/[critters-webpack-plugin]/dist/critters.js:254:39

  - critters.js:244 Critters.<anonymous>
    [webpack-tutorial-boilerplate]/[critters-webpack-plugin]/dist/critters.js:244:34
```

Here is the dependency liist
```
    "autoprefixer": "^9.0.1",
    "babel-core": "^6.26.3",
    "babel-loader": "^7.1.5",
    "babel-preset-env": "^1.7.0",
    "clean-webpack-plugin": "^0.1.19",
    "critters-webpack-plugin": "^1.1.0",
    "css-loader": "^1.0.0",
    "file-loader": "^1.1.11",
    "html-loader": "^0.5.5",
    "html-webpack-inline-svg-plugin": "^1.2.4",
    "html-webpack-plugin": "^3.2.0",
    "mini-css-extract-plugin": "^0.4.1",
    "node-sass": "^4.9.2",
    "optimize-css-assets-webpack-plugin": "^4.0.3",
    "postcss-loader": "^2.1.6",
    "purgecss-webpack-plugin": "^1.2.0",
    "sass-loader": "^7.0.3",
    "style-loader": "^0.21.0",
    "uglifyjs-webpack-plugin": "^1.2.7",
    "url-loader": "^1.0.1",
    "webpack": "^4.16.2",
    "webpack-cli": "^3.1.0",
    "webpack-merge": "^4.1.3"
```

### PR DESCRIPTION
**Pull request by prateekbh**
Fixes #9 and #6 

- brings latest css selector support

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -272,8 +272,12 @@ export default class Critters {
         rule.selectors = rule.selectors.filter(sel => {
           // Strip pseudo-elements and pseudo-classes, since we only care that their associated elements exist.
           // This means any selector for a pseudo-element or having a pseudo-class will be inlined if the rest of the selector matches.
-          sel = sel.replace(/::?(?:[a-z-]+)([.[#~&^:*]|\s|\n|$)/gi, '$1');
-          return document.querySelector(sel, document) != null;
+          sel = sel.replace(/::?(?:[a-z-]+)([.[#~&^:*]|\s|\n|$)/gi, '$1').trim();
+          try {
+            return document.querySelector(sel, document) != null;
+          } catch (e) {
+            return null;
```
Comment (prateekbh) [inline src/index.js]: @developit, I guess we can add a string telling which selector failed here

File: src/dom.js [hunk 2]
```diff
@@ -13,218 +13,27 @@
...
   return document;
 }
-
 /**
  * Serialize a Document to an HTML String
  * @param {Document} document   A Document, such as one created via `createDocument()`
  */
 export function serializeDocument (document) {
-  return parse5.serialize(document, PARSE5_OPTS);
+  return document.querySelector('html').innerHTML;
```
Comment (developit) [inline src/dom.js]: think this skips serialization of the doctype.
the JSDOM instance from line 24 has a `.serialize()` method that does this a little quicker and includes that info.

```js
  const jsdom = new JSDOM(html, {
    contentType: "text/html",
  });
  const { window } = jsdom;
  window.$jsdom = jsdom;
}
export function serializeDocument(document) {
  document.defaultView.$jsdom.serialize(document);
}
```

File: src/index.js [hunk 3]
```diff
@@ -252,7 +252,7 @@ export default class Critters {
     const head = document.querySelector('head');
 
     // basically `.textContent`
-    let sheet = style.childNodes.length > 0 && style.childNodes.map(node => node.nodeValue).join('\n');
+    let sheet = style.childNodes.length > 0 && [].slice.call(style.childNodes).map(node => node.nodeValue).join('\n');
```
Comment (developit) [inline src/index.js]: nit: `[].map.call(style.childNodes, ...`

File: src/index.js (hunk 1 above)
Comment (developit) [reply src/index.js]: that'd be good!

File: src/index.js [hunk 4]
```diff
@@ -272,8 +272,12 @@ export default class Critters {
         rule.selectors = rule.selectors.filter(sel => {
           // Strip pseudo-elements and pseudo-classes, since we only care that their associated elements exist.
           // This means any selector for a pseudo-element or having a pseudo-class will be inlined if the rest of the selector matches.
-          sel = sel.replace(/::?(?:[a-z-]+)([.[#~&^:*]|\s|\n|$)/gi, '$1');
-          return document.querySelector(sel, document) != null;
+          sel = sel.replace(/::?(?:[a-z-]+)([.[#~&^:*]|\s|\n|$)/gi, '$1').trim();
+          try {
+            return document.querySelector(sel, document) != null;
```
Comment (developit) [inline src/index.js]: I think we can drop the second argument here right?

**Comment by developit** [comment]
hush, sweet @googlebot

**Comment by developit** [comment]
Released as 1.2.0!

**Pull request merged**